sessiontxn/isolation/optimistic.go

```diff
@@ -110,7 +110,7 @@ func (p *OptimisticTxnContextProvider) GetStmtForUpdateTS() (uint64, error) {
 // AdviseOptimizeWithPlan providers optimization according to the plan
 // It will use MaxTS as the startTS in autocommit txn for some plans.
 func (p *OptimisticTxnContextProvider) AdviseOptimizeWithPlan(plan interface{}) (err error) {
-	if p.optimizeWithMaxTS || p.isTidbSnapshotEnabled() || p.isBeginStmtWithStaleRead() {
+	if p.optimizeWithMaxTS || p.isTidbSnapshotEnabled() || p.isBeginStmtWithStaleRead() || p.txn != nil {
```

Comment:
Why add this condition?

Reply:
`AdviseOptimizeWithPlan` is an optimization and should not have too many strong assumptions.  If the txn is already active, it's better to keep the statement goes well without error.

Reply:
It's more clear to use a method like `IsActive` to check whether it's already activated compared with `p.txn != nil`, maybe we could do some improvements in later PRs.

Reply:
Some comments added